osidb/sync_manager.py

```diff
@@ -328,11 +365,58 @@ def check_for_reschedules(cls):
                 )
                 continue
 
+    @classmethod
+    def is_in_progress(cls, sync_id):
+        """
+        Check if there is a task running for this sync manager. Task is considered running
+        if it has started and not finished yet or failed.
+        """
+
+        manager = cls.objects.get(sync_id=sync_id)
+
+        # This function is called before "last_started_dt" is set and
+        # a failed task has does not have "last_finished_dt" set
+        #
+        # 1) If "last_started_dt" is None, we assume that the task is NOT running.
+        # 2) If "last_started_dt" is not None, "last_failed_dt" is None, and "last_finished_dt" is None,
+        #    the task has started and is currently IN PROGRESS
+        # 3) If "last_started_dt" is not None, "last_failed_dt" is not None (and > last_started_dt),
+        #    and "last_finished_dt" is None, the previous run FAILED
+        # 4) If "last_started_dt" is not None and "last_finished_dt" is not None (and > last_started_dt),
+        #    the task has COMPLETED successfully
+
+        return (
+            # must have started
+            manager.last_started_dt is not None
+            and
+            # must NOT have failed since that start
+            not (
+                manager.last_failed_dt is not None
+                and manager.last_failed_dt > manager.last_started_dt
+            )
+            and
+            # and either not finished (-> running) or explicitly in-progress
+            manager.last_finished_dt is None
+        )
+
+    @classmethod
+    def is_scheduled(cls, sync_id):
+        """
+        Check if there's a task already scheduled to run in the future
+        """
+        manager = cls.objects.get(sync_id=sync_id)
+
+        return manager and (
+            manager.planned_start_dt is not None
+            and manager.planned_start_dt > timezone.now()
```

Comment:
I am not sure if I understand this logic. We set the `planned_start_dt`  ourselves, so what about the scenario when:
1. task is scheduled first time and `planned_start_dt` is set to e.g. `8:00` but it does not start because the queue is filled up with other tasks
2. meanwhile same task is scheduled in `8:01` for the second time and the first one didn't even start, so based on this logic, it will check whether the `planned_start_dt` (`8:00`) is greater than `now` (`8:01`), which should not be true and thus the `is_scheduled` will be evaluated to `False`

Reply:
@JakubFrejlach has good points.

I think that we should rather consider a task to be scheduled when it has `last_scheduled_dt` and it has either no `last_started_dt` or is is before `last_scheduled_dt`. I am not sure that `planned_start_dt` matters here as the start is not guaranteed to happen in that moment.

I am not sure this condition is sufficient. It would definitely not account for the cases of the tasks which were scheduled but failed to start for some reason - eg. Celery restart (hopefully rare).